File: clojure/aoc/src/aoc/y16/day01.clj
(ns aoc.y16.day01
  (:require [aoc.utils :as u]
            [clojure.string :as s]
            [clojure.core.match :refer [match]]))

(def input (u/read-input "y16/day01.txt"))

(defn parse-direction [instruction]
  (keyword (str (first instruction))))

(defn parse-distance [instruction]
  (parse-long (str (s/join (rest instruction)))))

(defn get-new-dir [currently-facing turn-direction]
  (match [currently-facing turn-direction]
    [:north :R] :east
    [:north :L] :west
    [:east :R] :south
    [:east :L] :north
    [:south :L] :east
    [:south :R] :west
    [:west :R] :north
    [:west :L] :south))

(defn move
  "Executes an instruction and returns an updated location based on what the instruction would do"
  [state instruction]
  ;; determine 
  (let [curdir (get state :facing)
        curloc (get state :location)
        turndir (parse-direction instruction)
        magnitude (parse-distance instruction)
        newdir (get-new-dir curdir turndir)]

    (cond
      (= newdir :north) {:facing newdir :location [(first curloc) (+ magnitude (second curloc))]}
      (= newdir :east) {:facing newdir :location [(+ magnitude (first curloc)) (second curloc)]}
      (= newdir :south) {:facing newdir :location [(first curloc) (- magnitude (second curloc))]}
      (= newdir :west)  {:facing newdir :location [(+ magnitude (first curloc)) (second curloc)]})))

(defn follow-instructions [] 
  (let [initial-state {:location [0 0] :facing :north}
        instructions (s/split input #", ")]
    (reduce (fn [state instruction] (move state instruction)) initial-state instructions))
  )

(defn part-1 []
  )

(u/manhattan-distance [0 0] (get ( follow-instructions) :location))


(follow-instructions)
(parse-direction "L54")
(parse-distance "L566")
(rest "L56")

File: aoc/y16/day01.py
with open('data/day01.txt', 'r') as f:
    directions = [d.strip() for d in f.read().split(',')]

cardinal_directions = {'N', 'S', 'E', 'W'}
turn_directions = {'L', 'R'}

def new_heading(current_heading, instruction_direction):
    assert current_heading in cardinal_directions
    assert instruction_direction in turn_directions

    heading_map = {
        'N': {'R': 'E', 'L': 'W'},
        'S': {'R': 'W', 'L': 'E'},
        'E': {'R': 'S', 'L': 'N'},
        'W': {'R': 'N', 'L': 'S'}
    }

    return heading_map[current_heading][instruction_direction]

def parse_dir(direction):
    turn_dir = direction[0]
    magnitude = int(direction[1:])

    return turn_dir, magnitude

current_heading = 'N'
location = (0, 0)

for direction in directions:
    turn_dir, distance = parse_dir(direction)
    current_heading = new_heading(current_heading, turn_dir)
    if current_heading == 'N':
        location = (location[0], location[1] + distance)
    elif current_heading == 'S':
        location = (location[0], location[1] - distance)
    elif current_heading == 'E':
        location = (location[0] + distance, location[1])
    elif current_heading == 'W':
        location = (location[0] - distance, location[1])

print(f'Location is {location}')
print(f'Manhattan distance is {abs(location[0] - 0) + abs(location[1] - 0)}')